safeweb/http.go

```diff
@@ -0,0 +1,217 @@
+			var h http.Handler
+			var apiRequest bool
+			h, p := browserHandler.Handler(r)
+			if p == "" {
+				h = apiHandler
+				apiRequest = true
+			}
+
+			if apiRequest {
+				switch r.Method {
+				// disallow `application/x-www-form-urlencoded` content-type for all API requests
+				case "POST", "PUT", "PATCH":
+					if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
+						http.Error(w, "invalid content type", http.StatusBadRequest)
+						return
+					}
+
+				// set CORS headers for pre-flight OPTIONS requests if any were configured
+				case "OPTIONS":
+					if len(config.PermittedCrossOriginHosts) > 0 {
+						w.Header().Set("Access-Control-Allow-Origin", strings.Join(config.PermittedCrossOriginHosts, ", "))
+						w.Header().Set("Access-Control-Allow-Methods", strings.Join(config.PermittedCrossOriginMethods, ", "))
+					}
+				}
+				h.ServeHTTP(w, r)
+			} else {
+				// TODO(@patrickod) consider templating additions to the CSP header.
+				w.Header().Set("Content-Security-Policy", defaultCSP)
+				w.Header().Set("X-Content-Type-Options", "nosniff")
+				csrfProtect(h).ServeHTTP(w, r)
+			}
+		})
+	}
+}
+
+func NewServer(config Config) (*Server, error) {
+	// ensure we have a valid listener configuration
+	if config.Hostname == "" && config.Listener == nil {
+		return nil, fmt.Errorf("must provide one of either Hostname or Listener")
+	}
+
+	// ensure that CORS configuration is complete
+	providedCORSMethods := len(config.PermittedCrossOriginMethods) > 0
+	providedCORSHosts := len(config.PermittedCrossOriginHosts) > 0
+	if providedCORSMethods != providedCORSHosts {
+		return nil, fmt.Errorf("must provide both PermittedCrossOriginMethods and PermittedCrossOriginHosts")
+	}
+
+	// fill in any missing fields
+	if err := config.setDefaults(); err != nil {
+		return nil, fmt.Errorf("failed to set defaults: %w", err)
+	}
+
+	return &Server{
+		config,
+		&http.Server{Handler: wrapHandler(config)(config.BrowserMux, config.APIMux)},
+	}, nil
+}
+
+func (s *Server) redirectHTTP(fqdn string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		u := r.URL
+		new := url.URL{
+			Scheme:   "https",
+			Host:     fqdn,
+			Path:     u.Path,
+			RawQuery: u.RawQuery,
+		}
+
+		http.Redirect(w, r, new.String(), http.StatusMovedPermanently)
+	}
+}
+
+func (s *Server) Serve() error {
+	// serve only HTTP if a listener is provided.
+	if s.Listener != nil {
+		return s.h.Serve(s.Listener)
+	}
+
+	// if a hostname is provided, create a tsnet server by that name and serve
+	// both HTTP and HTTPS traffic.
+	ts := tsnet.Server{
+		Hostname: s.Hostname,
+	}
+
+	// await a successful tsnet server up status to understand our FQDN for HTTP
+	// redirects.
+	var fqdn string
+up:
+	for {
+		upCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		status, err := ts.Up(upCtx)
+		if err == nil && status != nil {
```

Comment:
this can become an infinite fail-loop.
can we simply call `ts.Up` once without a timeout until it succeeds?